DESIGN.md

```diff
@@ -0,0 +1,225 @@
+# subiquity design notes
+
+## UI
+
+### basic ground rules:
+
+1. Subiquity is entirely usable by pressing up, down, space (or return) and the
+   occasional bit of typing.
+
+2. The UI never blocks.  If something takes more than about 0.1s, it is done in
+   the background, possibly with some kind of indication in the UI (and the
+   ability to cancel if appropriate).
```

Comment:
I wonder if we should state that showing progress pop-ups for small things results in too fast flicker. Maybe like progress popup should be a minimum of 0.5s, when shown. To prevent too fast flicker.

Reply:
Yeah, I wonder about this. Thinking about two examples where we block the UI like this (actually these might be the only two currently?):

 1. applying keyboard configuration -- I think for this we could make an argument for padding the runtime out to 0.5s if the subprocess finished quickly (but it always takes about 1.2s IME)
 2. downloading ssh keys -- I think here the case is a bit less clear, because what follows the download is still part of the same operation in some sense. But OTOH I guess padding out to 0.5s still isn't very long...

So um I dunno. I'll add a comment for now.